crates/lib3h/src/dht/mirror_dht.rs

```diff
@@ -0,0 +1,251 @@
+                None => break,
+                Some(msg) => msg,
+            };
+            let res = self.serve_DhtCommand(&cmd);
+            if let Ok(mut output) = res {
+                did_work = true;
+                outbox.append(&mut output);
+            } else {
+                println!("(log.e) serve_DhtCommand() failed: {:?}", res);
+            }
+        }
+        Ok((did_work, outbox))
+    }
+}
+
+/// Internals
+impl MirrorDht {
+    /// Return true if new peer or updated peer
+    fn add_peer(&mut self, peer_info: &PeerData) -> bool {
+        let maybe_peer = self.peer_list.get_mut(&peer_info.peer_address);
+        match maybe_peer {
+            None => {
+                self.peer_list
+                    .insert(peer_info.peer_address.clone(), peer_info.clone());
+                true
+            }
+            Some(mut peer) => {
+                if peer_info.timestamp <= peer.timestamp {
+                    return false;
+                }
+                peer.timestamp = peer_info.timestamp;
+                true
+            }
+        }
+    }
+
+    /// Add entry to our local storage
+    /// Return true if new entry was successfully added
+    fn add_entry(&mut self, entry: &EntryData) -> bool {
+        let maybe_entry = self.entry_list.get_mut(&entry.entry_address);
+        match maybe_entry {
+            None => {
+                self.entry_list
+                    .insert(entry.entry_address.clone(), entry.clone());
+                true
+            }
+            Some(prev_entry) => {
+                return prev_entry.merge(entry);
+            }
+        }
+    }
+
+    /// Process a DhtEvent Command, sent by our owner.
+    /// Return a list of DhtEvent to owner.
+    #[allow(non_snake_case)]
+    fn serve_DhtCommand(&mut self, cmd: &DhtCommand) -> Lib3hResult<Vec<DhtEvent>> {
+        println!("(log.d) --- '(MirrorDht)' serving cmd: {:?}", cmd);
+        // Note: use same order as the enum
+        match cmd {
+            // Received gossip from remote node. Bundle must be one of the following:
+            // - EntryData
+            // - PeerData
+            DhtCommand::HandleGossip(msg) => {
+                println!("Deserializer msg.bundle: {:?}", msg.bundle);
+                let mut de = Deserializer::new(&msg.bundle[..]);
+                let maybe_entry: Result<EntryData, rmp_serde::decode::Error> =
+                    Deserialize::deserialize(&mut de);
+                if let Ok(entry) = maybe_entry {
+                    let is_new = self.add_entry(&entry);
+                    if is_new {
+                        return Ok(vec![DhtEvent::HoldEntryRequested(entry)]);
+                    }
+                    return Ok(vec![]);
+                }
+                de = Deserializer::new(&msg.bundle[..]);
+                let maybe_peer: Result<PeerData, rmp_serde::decode::Error> =
+                    Deserialize::deserialize(&mut de);
```

Comment:
For the real dht implementation, we will need to land a well defined gossip and p2p messaging protocol. I'm cool with this one being more organic. But I think we should strive for code readability and debug-ability.

It might be easier to see what's going on here if we have a super enum, and can always reliably encode / decode it:

```rust
pub enum Gossip {
  Entry(EntryData),
  Peer(PeerData),
}
```

Reply:
yes, totally. Wanted to be sure I was going in the right direction before making this less sloppy :P